### PR TITLE
Dexdump

### PR DESCRIPTION
**Pull request by counter-reverse**
Solve this issue: https://github.com/Exodus-Privacy/exodus-core/issues/9 radically by removing the regex.  The regex used to get the wrong number of classes because these include types and superclasses. The unit test are also changed to get the real number of classes in a way that does not include types and superclasses.

Avoid any risk to get a command injection: https://github.com/Exodus-Privacy/exodus-core/issues/7

Solve this issue: https://github.com/Exodus-Privacy/exodus-core/issues/17 (no dexdump implies no check)

Make the code more readable: https://github.com/Exodus-Privacy/exodus-core/issues/16

### CODE REVIEW

File: exodus_core/analysis/static_analysis.py [hunk 1]
```diff
@@ -168,27 +168,17 @@ def get_embedded_classes(self):
...
-                raise Exception('Unable to decode the APK')
+        dexs = list(APK(filename=self.apk_path).get_all_dex())
+
+        class_names = []
+        for dex in dexs:
+            d = DalvikVMFormat(dex)
+            dex_class_names = d.get_classes_names()
+            for c in dex_class_names:
+                if "$" not in c:
+                    class_names.append(c[1:-1])
```
Comment (pnu-s) [inline exodus_core/analysis/static_analysis.py]: Not sure to understand the c[1:-1]

File: exodus_core/analysis/static_analysis.py [hunk 2]
```diff
@@ -168,27 +168,17 @@ def get_embedded_classes(self):
...
-            cmd = '{} {}/classes*.dex | perl -n -e\'/[A-Z]+((?:\w+\/)+\w+)/ && print "$1\n"\'|sort|uniq'.format(
-                dexdump, tmp_dir)
-            try:
-                self.classes = subprocess.check_output(
-                    cmd,
-                    stderr=subprocess.STDOUT,
-                    shell=True,
-                    universal_newlines=True
-                ).splitlines()
-                logging.debug('{} classes found in {}'.format(len(self.classes), self.apk_path))
```
Comment (pnu-s) [inline exodus_core/analysis/static_analysis.py:186]: We need to keep this debug log line somewhere

File: exodus_core/analysis/static_analysis.py [hunk 3]
```diff
@@ -168,27 +168,17 @@ def get_embedded_classes(self):
...
-                return self.classes
-            except subprocess.CalledProcessError:
-                logging.error('Unable to decode {}'.format(self.apk_path))
-                raise Exception('Unable to decode the APK')
+        dexs = list(APK(filename=self.apk_path).get_all_dex())
```
Comment (pnu-s) [inline exodus_core/analysis/static_analysis.py]: Don't we need to try/except somewhere in case Androguard raises an exception?

**Review by pnu-s — COMMENTED** [review]
The code changes look really cool if this works as expected.
I need to make tests with this new code to make sure it doesn't break anything

**Review by pnu-s — CHANGES_REQUESTED** [review]
I just made a few tests and there is a problem with some applications.
For instance this one: https://reports.exodus-privacy.eu.org/en/reports/128219/

With the new method, we do not detect 3 trackers:
* Facebook ads
* Millennial Media
* Twitter Mopub

File: exodus_core/analysis/static_analysis.py (hunk 1 above)
Comment (counter-reverse) [reply exodus_core/analysis/static_analysis.py]: @simpnu the name of the classes start with `L` and finish with `;`. For example `Lcom/google/common/cache/LocalCache;`. So I removed these characters to keep only the path and the name of the class.

**Comment by counter-reverse** [comment]
> I just made a few tests and there is a problem with some applications.
> For instance this one: https://reports.exodus-privacy.eu.org/en/reports/128219/
> 
> With the new method, we do not detect 3 trackers:
> 
>     * Facebook ads
> 
>     * Millennial Media
> 
>     * Twitter Mopub


I have manually opened the apk textra. Then I understand what is going on.

My current program get all classes defined in the apk. I believed all the imports are present in the apk. I was wrong. It probably only contains overided method classes.

I have modified my code to list all the classes that contain a method called in any class defined in the binary. My code now works and detects the 8 trackers of textra. I am going to push.

**Comment by counter-reverse** [comment]
> I just made a few tests and there is a problem with some applications.
> For instance this one: https://reports.exodus-privacy.eu.org/en/reports/128219/
> 
> With the new method, we do not detect 3 trackers:
> 
>     * Facebook ads
> 
>     * Millennial Media
> 
>     * Twitter Mopub


I have manually opened the apk textra. Then I understand what is going on.

My current program get all classes defined in the apk. I believed all the imports are present in the apk. I was wrong. It probably only contains overided method classes.

I have modified my code to list all the classes that contain a method called in any class defined in the binary. My code now works and detects the 8 trackers of textra. My code is still a bit slow. I am fixing that.

 I am going to push.

**Comment by gogo2464** [comment]
As told, I have changed the code. Now each call on each method definition of each class is notified. The 8 trackers of textra are corectly identified.

The code is also very slower now.

**Comment by counter-reverse** [comment]
@pnu-s I pushed. Now my code check each call in each class to find a tracker. It works. It is cleaner to read in the source code.

**Comment by counter-reverse** [comment]
I have checked if the tracker was called before that I add it to the tracker list. I think I should have done that in another branch. The software is now slower. If you want I can come back to the old version of my pull request with no check. It will be quicker.

**Comment by counter-reverse** [comment]
Some classses were missings. I have updated the unit tests on the classes. All is ok. We can merge!

File: setup.py [hunk 4]
```diff
@@ -31,11 +31,6 @@ def is_exe(fpath):
     print("Unfortunately, we do not support your platform %s" % sys.platform)
     sys.exit(1)
 
-if which('dexdump') is None:
-    print("Unable to find dexdump executable, please install it.")
-    print("On Debian-like OS, run sudo apt-get install dexdump")
-    sys.exit(1)
```
Comment (Gu1nness) [inline setup.py:37]: Since you removed dexdump and the call to `which` you can remove the function at the top of the file.

**Review by Gu1nness — CHANGES_REQUESTED** [review]
Please remove the `which` function at the top of the file.

**Comment by counter-reverse** [comment]
Nice.


I updated the unit test because the command used to add non-class object such as type and used to add it to the class list.

I also updated the unit test because the command used to ignore parent classes.


I think we can merge. :)

**Comment by counter-reverse** [comment]
I really would like you check my pull request.

**Comment by counter-reverse** [comment]
For the unit test I never knew what was the right number of classes. Too much to count and test. The current dexdump command takes more than it should.

But now I am sure to get the right number of class with the command `radare2 classes.dex; icq;` to list each. I will do it for each dex for each APK of the unit test and it should be nice.

**Comment by pnu-s** [comment]
@counter-reverse I'll take a look once more at this PR. I cannot guarantee when I will have time to do this though, but I'll try my best.

If you can resolve the conflict on the travis CI file that would be nice.

**Comment by counter-reverse** [comment]
Yes @pnu-s I will solve the conflict. Can you just tell me if everything is fine first please?

**Comment by counter-reverse** [comment]
@pnu-s I solved the conflict. Fell free to check.

**Comment by pnu-s** [comment]
As Androguard is not maintained anymore, I'm not sure we want to use even more of this tool and increase our dependency.

Closing this then.

**Comment by counter-reverse** [comment]
@pnu-s exodus core was already using androguard. 

I can only agree with the the fact it is a bad choice. 


Could you consider this PR as an incremental work please? 

Please accept this PR (after a rebase) and I will make a new one to migrate from androguard to lief for EVERY line of code of exodus privacy. Not only my pull request. 

I spent some full months to make this PR.

**Pull request closed without merging**